icu4c/source/tools/gendict/gendict.cpp

```diff
@@ -417,38 +423,49 @@ int  main(int argc, char **argv) {
         exit(status.reset());
     }
     if (verbose) { puts("Opening output file..."); }
-    UNewDataMemory *pData = udata_create(nullptr, nullptr, outFileName, &dataInfo, copyright, status);
-    if (status.isFailure()) {
-        fprintf(stderr, "gendict: could not open output file \"%s\", \"%s\"\n", outFileName, status.errorName());
-        exit(status.reset());
-    }
 
-    if (verbose) { puts("Writing to output file..."); }
-    int32_t indexes[DictionaryData::IX_COUNT] = {
-        DictionaryData::IX_COUNT * sizeof(int32_t), 0, 0, 0, 0, 0, 0, 0
-    };
-    int32_t size = outDataSize + indexes[DictionaryData::IX_STRING_TRIE_OFFSET];
-    indexes[DictionaryData::IX_RESERVED1_OFFSET] = size;
-    indexes[DictionaryData::IX_RESERVED2_OFFSET] = size;
-    indexes[DictionaryData::IX_TOTAL_SIZE] = size;
-
-    indexes[DictionaryData::IX_TRIE_TYPE] = isBytesTrie ? DictionaryData::TRIE_TYPE_BYTES : DictionaryData::TRIE_TYPE_UCHARS;
-    if (hasValues) {
-        indexes[DictionaryData::IX_TRIE_TYPE] |= DictionaryData::TRIE_HAS_VALUES;
-    }
+    if (isToml) {
+        FILE* f = fopen(outFileName, "w");
+        if (f == nullptr) {
+            fprintf(stderr, "gendict: could not open output file \"%s\"\n", outFileName);
+            exit(status.reset());
+        }
+        usrc_writeArray(f, "trie_data = [\n  ",  outData, isBytesTrie ? 8 : 16, outDataSize, "  ", "\n]\n");
```

Comment:
Observation: The old code writes both `indexes` and `outData` to the file, but this code writes only `outData`.

Reply:
Yeah because that was all that was being produced in the icuexportdata toml artefact. Could include indexes if someone tells me how it should be represented.

Reply:
I see, so it looks like the main thing `indexes` tells you is whether the trie bytes are from a BytesTrieBuilder or a UCharsTrieBuilder. I feel like that is important information to retain. If it wasn't being retained earlier that was just a limitation of the hacky CI script. If we upgrade the feature to be part of gendict itself, we should try to do the right thing.

It seems like the only useful information being written out is:

```
        indexes[DictionaryData::IX_TRIE_TYPE] = isBytesTrie ? DictionaryData::TRIE_TYPE_BYTES : DictionaryData::TRIE_TYPE_UCHARS;
        if (hasValues) {
            indexes[DictionaryData::IX_TRIE_TYPE] |= DictionaryData::TRIE_HAS_VALUES;
        }
        indexes[DictionaryData::IX_TRANSFORM] = dict.getTransform();
```

Proposed TOML structure:

```toml
trie_type = "bytes"  # or "uchars"
has_values = true  # or false
transform_type = "offset"  # or "none"
transform_offset = 0x123  # or maybe "transform_constant"? 0 if transform type is "none"
```

Reply:
Done.